File: code/cpp/quick_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

int partition(vector <int> &arr, int l, int r)
{
	int pivot = arr[r];

	int i = (l - 1);

	for (int j = l; j <= r; j++)
	{
		if (arr[j] <= pivot)
		{
			i++;
			swap(arr[i], arr[j]);
		}
	}

	return i;
}

void quick_sort(vector <int> &arr, int l, int r)
{
	if (l < r)
	{
		int p = partition(arr, l, r);

		quick_sort(arr, l, p - 1);
		quick_sort(arr, p + 1, r);
	}
}

int main()
{
	cout << "Enter number of elements: ";
	int n;	cin >> n;

	cout << "Array: ";
	vector <int> arr(n);

	for (int i = 0; i < n; i++)
		cin >> arr[i];

	quick_sort(arr, 0, n - 1);

	cout << "Sorted: ";

	for (auto u : arr)
		cout << u << " ";
	cout << "\n";

	return 0;
}

File: code/python/quicksort.py
def quick_sort(A, p, r):
    i=p
    j=r 
    x = A[int((p+r)/2)]
    while i <= j:
        while A[i] < x:
            i += 1
        while A[j] > x:
            j -= 1
        if i <= j: 
            A[i], A[j] = A[j], A[i]
            i += 1
            j -= 1
    if (p < j):
        quick_sort(A, p, j)
    if (r > i):
        quick_sort(A, i, r)
